Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/PDAnalyst.java

```diff
@@ -669,34 +671,133 @@ public void updateLog(PLog pLog, APMLog apmLog) throws Exception {
-                    .filter(
-                        activityInstance -> activityInstance.getAttributes().containsKey(Constants.ATT_KEY_RESOURCE))
-                    .map(activityInstance -> activityInstance.getAttributes().get(Constants.ATT_KEY_RESOURCE))
-                    .collect(Collectors.toSet())
-                    .size())
+                .resourceCountByRole(getResourceCountsGroupedByRole())
+                .resourceCount(getTotalResourcesCount())
+                .nodeIdToRoleName(getRoleForNodeId(bpmnAbstraction))
                 .startTime(filteredAPMLog.getStartTime())
                 .endTime(filteredAPMLog.getEndTime())
-                .nodeWeights(nodeDurationMap)
+                .nodeWeights(getNodeWeights(bpmnAbstraction))
                 .edgeFrequencies(groupOutboundEdgeFrequenciesByGateway(bpmnAbstraction))
                 .build();
         }
         return simulationData;
     }
 
+    private Map<String, Double> getNodeWeights(BPMNAbstraction bpmnAbstraction) {
+        return bpmnAbstraction.getDiagram().getNodes()
+            .stream()
+            .filter(Activity.class::isInstance)
+            .collect(Collectors.toMap(
+                node -> node.getId().toString(),
+                node -> attLog.getGraphView().getNodeWeight(node.getLabel(),
+                        MeasureType.DURATION,
+                        MeasureAggregation.MEAN,
+                        MeasureRelation.ABSOLUTE)));
+    }
+
+    private long getTotalResourcesCount() {
+        return filteredPLog.getActivityInstances().stream()
+            .filter(
+                activityInstance -> activityInstance.getAttributes().containsKey(Constants.ATT_KEY_RESOURCE))
+            .map(activityInstance -> activityInstance.getAttributes().get(Constants.ATT_KEY_RESOURCE))
+            .collect(Collectors.toSet())
+            .size();
+    }
+
+    private Map<String, Integer> getResourceCountsGroupedByRole() {
+        Map<String, Set<String>> rolesToResourcesMap = new HashMap<>();
+        filteredPLog.getActivityInstances().stream()
+            .filter(activityInstance -> activityInstance.getAttributes().containsKey(Constants.ATT_KEY_RESOURCE))
+            .forEach(activityInstance -> {
+                String role = activityInstance.getAttributeValue(Constants.ATT_KEY_ROLE);
+
+                if (role == null || role.isEmpty()) {
+                    role = SimulationData.DEFAULT_ROLE;
+                }
+                Set<String> resources = rolesToResourcesMap.get(role);
+                if (resources == null) {
+                    resources = new HashSet<>();
+                }
+                resources.add(activityInstance.getResource());
+                rolesToResourcesMap.put(role, resources);
+            });
+
+        return rolesToResourcesMap.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, stringSetEntry -> stringSetEntry.getValue().size()));
```

Comment:
What happens when activity instances could have org:role attribute (e.g. Role1) but no org:resource, Role1 should be mapped to DEFAULT_RESOURCE but it is omitted.
Maybe we should make it consistent rule: for any activity instance, no org:role is considered as DEFAULT_ROLE, no org:resource is considered as DEFAULT_RESOURCE and it could be helpful to avoid if role is not DEFAULT_ROLE then...else in other code that uses this mapping?

If the rule is consistent above, we can make the code easy:
- No need to filter out activity instances without org:resource attribute.
- resource = activityInstance.getAttributes().containsKey(Constants.ATT_KEY_RESOURCE) ? activityInstance.getResource() : DEFAULT_RESOURCE
- role = activityInstance.getAttributes().containsKey(Constants.ATT_KEY_ROLE) ? activityInstance.getAttributeValue(Constants.ATT_KEY_ROLE) : DEFAULT_ROLE
- Add resource to the set of role's resources.
- No need if (role is DEFAULT_ROLE) then....else.... in other code.